cts/cts-regression.in

```diff
@@ -1,230 +1,257 @@
+    UNSAFE = 107
+    EXISTS = 108
+    MULTIPLE = 109
+    EXPIRED = 110
+    NOT_YET_IN_EFFECT = 111
+    INDETERMINATE = 112
+    UNSATISFIED = 113
+
+    # Other
+    TIMEOUT = 124
+
+    # OCF Resource Agent API 1.1
+    DEGRADED = 190
+    DEGRADED_PROMOTED = 191
+
+    # Custom
+    NONE = 193
+
+    MAX = 255
+
+
+class TestsArgAction(argparse.Action):
+    """Handle special cases for `tests` arguments.
+
+    Enforce a default value of ['cli', 'scheduler'].
+    Replace the 'all' alias with the tests that it represents.
+    Get rid of duplicates.
+
+    The main motivation is that when the `choices` argument of
+    :meth:`parser.add_argument()` is specified, the `default` argument
+    must contain exactly one value (not `None` and not a list). We want
+    our default to be a list of tests, namely `cli` and `scheduler`.
+    """
+
+    def __call__(self, parser, namespace, values, option_string=None):
+        all_tests = ['cli', 'exec', 'fencing', 'scheduler']
+        default_tests = ['cli', 'scheduler']
+
+        if not values:
+            setattr(namespace, self.dest, default_tests)
+            return
+
+        # If no argument is specified, the default gets passed as a
+        # string 'default' instead of as a list ['default']. Probably
+        # a bug in argparse. The below gives us a list.
+        if not isinstance(values, list):
+            values = [values]
+
+        tests = set(values)
+
+        # If 'all', is found, replace it with the tests it represents.
+        try:
+            tests.remove('all')
+            tests.update(set(all_tests))
+        except KeyError:
+            pass
+
+        # Same for 'default'
+        try:
+            tests.remove('default')
+            tests.update(set(default_tests))
+        except KeyError:
+            pass
+
+        setattr(namespace, self.dest, sorted(list(tests)))
+
+
+def error_print(msg):
+    """Print an error message.
+
+    :param msg: Message to print
+    :type msg: str
+    """
+    print('      * ERROR:   %s' % msg)
+
+
+def run_test(test, test_home, verbose=False, valgrind=False):
+    """Run a regression test and return the result.
+
+    :param test: The name of the test to run
+    :type test: str
+    :return: The exit code of the test
+    :rtype: :class:`CrmExit`
+    """
+    print('Executing the %s regression tests' % test)
+    print('=' * 60)
+    if test == 'pacemaker_remote':
+        cmd = [os.path.join(test_home, 'cts-exec'), '-R']
+    else:
+        cmd = [os.path.join(test_home, 'cts-%s' % test)]
+
+    if verbose:
+        cmd.append('--verbose')
+
+    # Valgrind is supported only for cli and scheduler currently
+    if valgrind and test in ['cli', 'scheduler']:
+        cmd.append('--valgrind')
+
+    if os.access(cmd[0], os.X_OK):
+        # Test might not be executable if run from source directory
+        mode = os.stat(cmd[0]).st_mode
+        os.chmod(cmd[0], mode | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH)
```

Comment:
It's not at all clear to me that this is needed or desirable. However, `chmod a+x` was in the bash version, and this is the Python equivalent.

Reply:
Now that the individual tests are all .in files, and properly configured in ./configure to be executable, it's not needed. There was a time when that wasn't true for all tests and this was useful.

Reply:
Gotcha, removed.